frontend/src/app/utils/api.ts

```diff
@@ -39,19 +40,30 @@ class FetchClient {
     ).href;
   };
 
-  getOptions = (body: JsonObject | null): RequestInit => {
+  getOptions = (
+    method: RequestMethod,
+    body: JsonObject | null
+  ): RequestInit => {
     return {
       ...this.defaultOptions,
-      method: "POST",
+      method,
       headers,
       body: body ? JSON.stringify(body) : undefined,
     };
   };
 
-  request = async (path: string, body: JsonObject | null) => {
-    const res = await fetch(this.getURL(path), this.getOptions(body));
+  request = async (
+    path: string,
+    body: JsonObject | null = null,
+    method: RequestMethod = "POST",
+    query = ""
+  ) => {
+    const res = await fetch(
+      this.getURL(path + query),
+      this.getOptions(method, body)
+    );
     if (!res.ok) {
-      throw res;
+      throw await res.json();
```

Comment:
**question:** If `.json` throws for some reason, will this result in an unhandled promise rejection, escaping the explicit `throw` and breaking any upstream code?

Reply:
Good call! I wrapped this in another try/catch.

Reply:
A little update here, I tried wrapping this in a try/catch but then it always displayed the default error message instead of the message in `res.json()`. I don't remember all the specifics, but this does remind me of an issue Nick ran into when he was writing this -  `res.json()` returns a Promise, so I think we had an issue with the fact that it jumped to the `catch` block too quickly, before waiting for the promise to resolve. 

Any ideas on what to do about this? I'm not sure how often we can expect `.json` to throw.

Reply:
I'm wondering if this could work?
```suggestion
     throw await res.json().catch((error) => error);
```